File: mRemoteNG/UI/UI.Window.ExternalTools.vb
Imports mRemoteNG.App
Imports WeifenLuo.WinFormsUI.Docking
Imports mRemoteNG.App.Runtime
Imports mRemoteNG.My

Namespace UI
    Namespace Window
        Public Class ExternalTools
            Inherits Base
#Region "Constructors"
            Public Sub New(ByVal panel As DockContent)
                InitializeComponent()

                WindowType = Type.ExternalApps
                DockPnl = panel
            End Sub
#End Region

#Region "Private Fields"
            Private _selectedTool As Tools.ExternalTool = Nothing
#End Region

#Region "Private Methods"
#Region "Event Handlers"
            Private Sub ExternalTools_Load(ByVal sender As Object, ByVal e As EventArgs) Handles Me.Load
                ApplyLanguage()
                UpdateToolsListView()
            End Sub

            Private Shared Sub ExternalTools_FormClosed(sender As System.Object, e As FormClosedEventArgs) Handles MyBase.FormClosed
                mRemoteNG.Config.Settings.Save.SaveExternalAppsToXML()
            End Sub

            Private Sub NewTool_Click(ByVal sender As System.Object, ByVal e As EventArgs) Handles NewToolMenuItem.Click, NewToolToolstripButton.Click
                Try
                    Dim externalTool As New Tools.ExternalTool(Language.strExternalToolDefaultName)
                    Runtime.ExternalTools.Add(externalTool)
                    UpdateToolsListView(externalTool)
                    DisplayNameTextBox.Focus()
                Catch ex As Exception
                    MessageCollector.AddExceptionMessage("UI.Window.ExternalTools.NewTool_Click() failed.", ex, , True)
                End Try
            End Sub

            Private Sub DeleteTool_Click(ByVal sender As System.Object, ByVal e As EventArgs) Handles DeleteToolMenuItem.Click, DeleteToolToolstripButton.Click
                Try
                    Dim message As String
                    Select Case ToolsListView.SelectedItems.Count
                        Case Is = 1
                            message = String.Format(Language.strConfirmDeleteExternalTool, ToolsListView.SelectedItems(0).Text)
                        Case Is > 1
                            message = String.Format(Language.strConfirmDeleteExternalToolMultiple, ToolsListView.SelectedItems.Count)
                        Case Else
                            Return
                    End Select

                    If Not MsgBox(message, MsgBoxStyle.Question Or MsgBoxStyle.YesNo) = MsgBoxResult.Yes Then Return

                    For Each listViewItem As ListViewItem In ToolsListView.SelectedItems
                        Dim externalTool As Tools.ExternalTool = TryCast(listViewItem.Tag, Tools.ExternalTool)
                        If externalTool Is Nothing Then Continue For

                        Runtime.ExternalTools.Remove(listViewItem.Tag)
                        listViewItem.Remove()
                    Next
                Catch ex As Exception
                    MessageCollector.AddExceptionMessage("UI.Window.ExternalTools.DeleteTool_Click() failed.", ex, , True)
                End Try
            End Sub

            Private Sub LaunchTool_Click(ByVal sender As System.Object, ByVal e As EventArgs) Handles LaunchToolMenuItem.Click, LaunchToolToolstripButton.Click
                LaunchTool()
            End Sub

            Private Sub ToolsListView_SelectedIndexChanged(ByVal sender As System.Object, ByVal e As EventArgs) Handles ToolsListView.SelectedIndexChanged
                Try
                    If ToolsListView.SelectedItems.Count = 1 Then
                        PropertiesGroupBox.Enabled = True
                        _selectedTool = TryCast(ToolsListView.SelectedItems(0).Tag, Tools.ExternalTool)
                        If _selectedTool Is Nothing Then Return

                        With _selectedTool
                            DisplayNameTextBox.Text = .DisplayName
                            FilenameTextBox.Text = .FileName
                            ArgumentsCheckBox.Text = .Arguments
                            WaitForExitCheckBox.Checked = .WaitForExit
                            TryToIntegrateCheckBox.Checked = .TryIntegrate
                        End With
                    Else
                        PropertiesGroupBox.Enabled = False
                    End If
                Catch ex As Exception
                    MessageCollector.AddExceptionMessage("UI.Window.ExternalTools.ToolsListView_SelectedIndexChanged() failed.", ex, , True)
                End Try
            End Sub

            Private Sub ToolsListView_DoubleClick(ByVal sender As Object, ByVal e As EventArgs) Handles ToolsListView.DoubleClick
                If ToolsListView.SelectedItems.Count > 0 Then LaunchTool()
            End Sub

            Private Sub PropertyControl_ChangedOrLostFocus(ByVal sender As Object, ByVal e As EventArgs) Handles DisplayNameTextBox.LostFocus, ArgumentsCheckBox.LostFocus, FilenameTextBox.LostFocus, BrowseButton.LostFocus, WaitForExitCheckBox.LostFocus, WaitForExitCheckBox.Click, TryToIntegrateCheckBox.Click
                If _selectedTool Is Nothing Then Return

                Try
                    With _selectedTool
                        .DisplayName = DisplayNameTextBox.Text
                        .FileName = FilenameTextBox.Text
                        .Arguments = ArgumentsCheckBox.Text
                        .WaitForExit = WaitForExitCheckBox.Checked
                        .TryIntegrate = TryToIntegrateCheckBox.Checked
                    End With

                    UpdateToolsListView()
                Catch ex As Exception
                    MessageCollector.AddExceptionMessage("UI.Window.ExternalTools.PropertyControl_ChangedOrLostFocus() failed.", ex, , True)
                End Try
            End Sub

            Private Sub BrowseButton_Click(ByVal sender As System.Object, ByVal e As EventArgs) Handles BrowseButton.Click
                Try
                    Using browseDialog As New OpenFileDialog()
                        With browseDialog
                            .Filter = String.Join("|", New String() {Language.strFilterApplication, "*.exe", Language.strFilterAll, "*.*"})
                            If .ShowDialog = DialogResult.OK Then FilenameTextBox.Text = .FileName
                        End With
                    End Using
                Catch ex As Exception
                    MessageCollector.AddExceptionMessage("UI.Window.ExternalTools.BrowseButton_Click() failed.", ex, , True)
                End Try
            End Sub

            Private Sub TryToIntegrateCheckBox_CheckedChanged(ByVal sender As System.Object, ByVal e As EventArgs) Handles TryToIntegrateCheckBox.CheckedChanged
                If TryToIntegrateCheckBox.Checked Then
                    WaitForExitCheckBox.Enabled = False
                    WaitForExitCheckBox.Checked = False
                Else
                    WaitForExitCheckBox.Enabled = True
                End If
            End Sub
#End Region

            Private Sub ApplyLanguage()
                Text = Language.strMenuExternalTools
                TabText = Language.strMenuExternalTools

                NewToolToolstripButton.Text = Language.strButtonNew
                DeleteToolToolstripButton.Text = Language.strOptionsKeyboardButtonDelete
                LaunchToolToolstripButton.Text = Language.strButtonLaunch

                DisplayNameColumnHeader.Text = Language.strColumnDisplayName
                FilenameColumnHeader.Text = Language.strColumnFilename
                ArgumentsColumnHeader.Text = Language.strColumnArguments
                WaitForExitColumnHeader.Text = Language.strColumnWaitForExit
                TryToIntegrateCheckBox.Text = Language.strTryIntegrate

                PropertiesGroupBox.Text = Language.strGroupboxExternalToolProperties

                DisplayNameLabel.Text = Language.strLabelDisplayName
                FilenameLabel.Text = Language.strLabelFilename
                ArgumentsLabel.Text = Language.strLabelArguments
                OptionsLabel.Text = Language.strLabelOptions
                WaitForExitCheckBox.Text = Language.strCheckboxWaitForExit
                BrowseButton.Text = Language.strButtonBrowse

                NewToolMenuItem.Text = Language.strMenuNewExternalTool
                DeleteToolMenuItem.Text = Language.strMenuDeleteExternalTool
                LaunchToolMenuItem.Text = Language.strMenuLaunchExternalTool
            End Sub

            Private Sub UpdateToolsListView(Optional ByVal selectTool As Tools.ExternalTool = Nothing)
                Try
                    Dim selectedTools As New List(Of Tools.ExternalTool)
                    If selectTool Is Nothing Then
                        For Each listViewItem As ListViewItem In ToolsListView.SelectedItems
                            Dim externalTool As Tools.ExternalTool = TryCast(listViewItem.Tag, Tools.ExternalTool)
                            If externalTool IsNot Nothing Then selectedTools.Add(externalTool)
                        Next
                    Else
                        selectedTools.Add(selectTool)
                    End If

                    ToolsListView.BeginUpdate()
                    ToolsListView.Items.Clear()

                    For Each externalTool As Tools.ExternalTool In Runtime.ExternalTools
                        Dim listViewItem As New ListViewItem
                        With listViewItem
                            .Text = externalTool.DisplayName
                            .SubItems.Add(externalTool.FileName)
                            .SubItems.Add(externalTool.Arguments)
                            .SubItems.Add(externalTool.WaitForExit)
                            .SubItems.Add(externalTool.TryIntegrate)
                            .Tag = externalTool
                        End With

                        ToolsListView.Items.Add(listViewItem)

                        If selectedTools.Contains(externalTool) Then listViewItem.Selected = True
                    Next

                    ToolsListView.EndUpdate()

                    frmMain.AddExternalToolsToToolBar()
                Catch ex As Exception
                    MessageCollector.AddExceptionMessage("UI.Window.ExternalTools.PopulateToolsListView()", ex, , True)
                End Try
            End Sub

            Private Sub LaunchTool()
                Try
                    For Each listViewItem As ListViewItem In ToolsListView.SelectedItems
                        Dim externalTool As Tools.ExternalTool = TryCast(listViewItem.Tag, Tools.ExternalTool)
                        If externalTool Is Nothing Then Continue For

                        externalTool.Start()
                    Next
                Catch ex As Exception
                    MessageCollector.AddExceptionMessage("UI.Window.ExternalTools.LaunchTool() failed.", ex, , True)
                End Try
            End Sub
#End Region
        End Class
    End Namespace
End Namespace

File: mRemoteV1.ConvertedToC#/UI/UI.Window.ExternalTools.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Drawing;
using System.Diagnostics;
using System.Windows.Forms;
using mRemoteNG.App;
using WeifenLuo.WinFormsUI.Docking;
using mRemoteNG.App.Runtime;
using mRemoteNG.My;

namespace mRemoteNG.UI
{
	namespace Window
	{
		public partial class ExternalTools : Base
		{
			#region "Constructors"
			public ExternalTools(DockContent panel)
			{
				FormClosed += ExternalTools_FormClosed;
				Load += ExternalTools_Load;
				InitializeComponent();

				WindowType = Type.ExternalApps;
				DockPnl = panel;
			}
			#endregion

			#region "Private Fields"
				#endregion
			private Tools.ExternalTool _selectedTool = null;

			#region "Private Methods"
			#region "Event Handlers"
			private void ExternalTools_Load(object sender, EventArgs e)
			{
				ApplyLanguage();
				UpdateToolsListView();
			}

			private static void ExternalTools_FormClosed(System.Object sender, FormClosedEventArgs e)
			{
				mRemoteNG.Config.Settings.Save.SaveExternalAppsToXML();
			}

			private void NewTool_Click(System.Object sender, EventArgs e)
			{
				try {
					Tools.ExternalTool externalTool = new Tools.ExternalTool(Language.strExternalToolDefaultName);
					Runtime.ExternalTools.Add(externalTool);
					UpdateToolsListView(externalTool);
					DisplayNameTextBox.Focus();
				} catch (Exception ex) {
					Runtime.MessageCollector.AddExceptionMessage("UI.Window.ExternalTools.NewTool_Click() failed.", ex, , true);
				}
			}

			private void DeleteTool_Click(System.Object sender, EventArgs e)
			{
				try {
					string message = null;
					switch (ToolsListView.SelectedItems.Count) {
						case  // ERROR: Case labels with binary operators are unsupported : Equality
1:
							message = string.Format(Language.strConfirmDeleteExternalTool, ToolsListView.SelectedItems[0].Text);
							break;
						case  // ERROR: Case labels with binary operators are unsupported : GreaterThan
1:
							message = string.Format(Language.strConfirmDeleteExternalToolMultiple, ToolsListView.SelectedItems.Count);
							break;
						default:
							return;
					}

					if (!(Interaction.MsgBox(message, MsgBoxStyle.Question | MsgBoxStyle.YesNo) == MsgBoxResult.Yes))
						return;

					foreach (ListViewItem listViewItem in ToolsListView.SelectedItems) {
						Tools.ExternalTool externalTool = listViewItem.Tag as Tools.ExternalTool;
						if (externalTool == null)
							continue;

						Runtime.ExternalTools.Remove(listViewItem.Tag);
						listViewItem.Remove();
					}
				} catch (Exception ex) {
					Runtime.MessageCollector.AddExceptionMessage("UI.Window.ExternalTools.DeleteTool_Click() failed.", ex, , true);
				}
			}

			private void LaunchTool_Click(System.Object sender, EventArgs e)
			{
				LaunchTool();
			}

			private void ToolsListView_SelectedIndexChanged(System.Object sender, EventArgs e)
			{
				try {
					if (ToolsListView.SelectedItems.Count == 1) {
						PropertiesGroupBox.Enabled = true;
						_selectedTool = ToolsListView.SelectedItems[0].Tag as Tools.ExternalTool;
						if (_selectedTool == null)
							return;

						var _with1 = _selectedTool;
						DisplayNameTextBox.Text = _with1.DisplayName;
						FilenameTextBox.Text = _with1.FileName;
						ArgumentsCheckBox.Text = _with1.Arguments;
						WaitForExitCheckBox.Checked = _with1.WaitForExit;
						TryToIntegrateCheckBox.Checked = _with1.TryIntegrate;
					} else {
						PropertiesGroupBox.Enabled = false;
					}
				} catch (Exception ex) {
					Runtime.MessageCollector.AddExceptionMessage("UI.Window.ExternalTools.ToolsListView_SelectedIndexChanged() failed.", ex, , true);
				}
			}

			private void ToolsListView_DoubleClick(object sender, EventArgs e)
			{
				if (ToolsListView.SelectedItems.Count > 0)
					LaunchTool();
			}

			private void PropertyControl_ChangedOrLostFocus(object sender, EventArgs e)
			{
				if (_selectedTool == null)
					return;

				try {
					var _with2 = _selectedTool;
					_with2.DisplayName = DisplayNameTextBox.Text;
					_with2.FileName = FilenameTextBox.Text;
					_with2.Arguments = ArgumentsCheckBox.Text;
					_with2.WaitForExit = WaitForExitCheckBox.Checked;
					_with2.TryIntegrate = TryToIntegrateCheckBox.Checked;

					UpdateToolsListView();
				} catch (Exception ex) {
					Runtime.MessageCollector.AddExceptionMessage("UI.Window.ExternalTools.PropertyControl_ChangedOrLostFocus() failed.", ex, , true);
				}
			}

			private void BrowseButton_Click(System.Object sender, EventArgs e)
			{
				try {
					using (OpenFileDialog browseDialog = new OpenFileDialog()) {
						var _with3 = browseDialog;
						_with3.Filter = string.Join("|", new string[] {
							Language.strFilterApplication,
							"*.exe",
							Language.strFilterAll,
							"*.*"
						});
						if (_with3.ShowDialog == DialogResult.OK)
							FilenameTextBox.Text = _with3.FileName;
					}
				} catch (Exception ex) {
					Runtime.MessageCollector.AddExceptionMessage("UI.Window.ExternalTools.BrowseButton_Click() failed.", ex, , true);
				}
			}

			private void TryToIntegrateCheckBox_CheckedChanged(System.Object sender, EventArgs e)
			{
				if (TryToIntegrateCheckBox.Checked) {
					WaitForExitCheckBox.Enabled = false;
					WaitForExitCheckBox.Checked = false;
				} else {
					WaitForExitCheckBox.Enabled = true;
				}
			}
			#endregion

			private void ApplyLanguage()
			{
				Text = Language.strMenuExternalTools;
				TabText = Language.strMenuExternalTools;

				NewToolToolstripButton.Text = Language.strButtonNew;
				DeleteToolToolstripButton.Text = Language.strOptionsKeyboardButtonDelete;
				LaunchToolToolstripButton.Text = Language.strButtonLaunch;

				DisplayNameColumnHeader.Text = Language.strColumnDisplayName;
				FilenameColumnHeader.Text = Language.strColumnFilename;
				ArgumentsColumnHeader.Text = Language.strColumnArguments;
				WaitForExitColumnHeader.Text = Language.strColumnWaitForExit;
				TryToIntegrateCheckBox.Text = Language.strTryIntegrate;

				PropertiesGroupBox.Text = Language.strGroupboxExternalToolProperties;

				DisplayNameLabel.Text = Language.strLabelDisplayName;
				FilenameLabel.Text = Language.strLabelFilename;
				ArgumentsLabel.Text = Language.strLabelArguments;
				OptionsLabel.Text = Language.strLabelOptions;
				WaitForExitCheckBox.Text = Language.strCheckboxWaitForExit;
				BrowseButton.Text = Language.strButtonBrowse;

				NewToolMenuItem.Text = Language.strMenuNewExternalTool;
				DeleteToolMenuItem.Text = Language.strMenuDeleteExternalTool;
				LaunchToolMenuItem.Text = Language.strMenuLaunchExternalTool;
			}

			private void UpdateToolsListView(Tools.ExternalTool selectTool = null)
			{
				try {
					List<Tools.ExternalTool> selectedTools = new List<Tools.ExternalTool>();
					if (selectTool == null) {
						foreach (ListViewItem listViewItem in ToolsListView.SelectedItems) {
							Tools.ExternalTool externalTool = listViewItem.Tag as Tools.ExternalTool;
							if (externalTool != null)
								selectedTools.Add(externalTool);
						}
					} else {
						selectedTools.Add(selectTool);
					}

					ToolsListView.BeginUpdate();
					ToolsListView.Items.Clear();

					foreach (Tools.ExternalTool externalTool in Runtime.ExternalTools) {
						ListViewItem listViewItem = new ListViewItem();
						var _with4 = listViewItem;
						_with4.Text = externalTool.DisplayName;
						_with4.SubItems.Add(externalTool.FileName);
						_with4.SubItems.Add(externalTool.Arguments);
						_with4.SubItems.Add(externalTool.WaitForExit);
						_with4.SubItems.Add(externalTool.TryIntegrate);
						_with4.Tag = externalTool;

						ToolsListView.Items.Add(listViewItem);

						if (selectedTools.Contains(externalTool))
							listViewItem.Selected = true;
					}

					ToolsListView.EndUpdate();

					My.MyProject.Forms.frmMain.AddExternalToolsToToolBar();
				} catch (Exception ex) {
					Runtime.MessageCollector.AddExceptionMessage("UI.Window.ExternalTools.PopulateToolsListView()", ex, , true);
				}
			}

			private void LaunchTool()
			{
				try {
					foreach (ListViewItem listViewItem in ToolsListView.SelectedItems) {
						Tools.ExternalTool externalTool = listViewItem.Tag as Tools.ExternalTool;
						if (externalTool == null)
							continue;

						externalTool.Start();
					}
				} catch (Exception ex) {
					Runtime.MessageCollector.AddExceptionMessage("UI.Window.ExternalTools.LaunchTool() failed.", ex, , true);
				}
			}
			#endregion
		}
	}
}
